features/123-chat-push-notification/api.md

```diff
@@ -0,0 +1,158 @@
+# Push Notification in Chat
+## Feature Overview
+
+This feature enables developer to send customized push notification via cloud function deployment. The approach is to provide hooks in chat conversation including
+
+* Messages
+  * after\_message\_sent
+  * after\_message\_updated
+  * after\_message\_deleted
+* Typing
+  * typing\_started
+* Conversation
+  * after\_conversation\_created
+  * after\_conversation\_deleted
+  * after\_conversation\_updated
+  * after\_users\_entered\_conversation
+  * after\_users\_leave\_conversation
+
+In addition, the current JS push notification code snippet is too verbose and can be simplified.
+
+### Task
+- Refactor JS SDK Push Notification Function so that it can be called by developer with less code.
+	- Original Code:
+
+	```javascript
+	var skygear = require('skygear');
+	skygear.push.sendToDevice(
+	  ['2aa4af2a-699a-4e43-8d67-7598757fc7ed'], // Device IDs
+	  {
+	    'apns': {
+	        'aps': {
+	            'alert': {
+	                'title': title,
+	                'body': message,
+	            }
+	        },
+	        'from': 'skygear',
+	        'operation': 'notification',
+	    },
+	    'gcm': {
+	         'notification': {
+	              'title': title,
+	              'body': message,
+	          }
+	    },
+		}
+	);
+	```
+
+	- new Code:
+
+	```javascript
+	var skygear = require('skygear');
+	skygear.push.sendToUser(
+	  [userBen, userRick],
+	  {
+         'title': title,
+         'body': message
+	  }
+	);
+	```
+- Create chat hook annotation in chat SDK
+- Enable chat plugin to call pre-defined hook
+
+### Sample Codes
+
+- Sample Javascript Cloud Function
+
+```javascript
+skygearChat.afterMessageSent((message, conversation) => {
+  skygear.push.sendToUser(conversation.participants, {'title': 'Title', 'body': 'Hello World'})
```

Comment:
Better to enrich the js example like py version, instead of static title and body